collections/deep_merge.ts

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Merges the two given Records, recursively merging any nested Records with
+ * the second collection overriding the first in case of conflict
+ *
+ * For arrays, maps and sets, a merging strategy can be specified to either
+ * "replace" values, or "concat" them instead.
+ * Use "includeNonEnumerable" option to include non enumerable properties too.
+ */
+export function deepMerge<
+  T extends Record<PropertyKey, unknown>,
+  U extends Record<PropertyKey, unknown>,
+>(
+  collection: T,
+  other: U,
+  {
+    arrays = "replace",
+    maps = "replace",
+    sets = "replace",
+    includeNonEnumerable = false,
+  }: deepMergeOptions = {},
+): T & U {
+  // Extract property and symbols
+  const keys = [
+    ...Object.getOwnPropertyNames(other),
+    ...Object.getOwnPropertySymbols(other),
+  ].filter((key) => includeNonEnumerable || other.propertyIsEnumerable(key));
+
+  // Iterate through each key of other object and use correct merging strategy
+  for (const key of keys as propertyKey[]) {
+    const a = collection[key], b = other[key];
+
+    // Handle arrays
+    if ((Array.isArray(a)) && (Array.isArray(b))) {
+      if (arrays === "concat") {
+        collection[key] = a.concat(...b);
```

Comment:
You create an intermediate array here which needs to be created and collected - I would suggest doing this in place (mutating) instead:

```suggestion
        collection[key] = a.push(...b);
```